crates/services/p2p/src/peer_report.rs

```diff
@@ -0,0 +1,452 @@
+use crate::{
+    config::Config,
+    heartbeat::{
+        Heartbeat,
+        HeartbeatEvent,
+    },
+};
+use fuel_core_types::fuel_types::BlockHeight;
+use libp2p::{
+    core::{
+        connection::ConnectionId,
+        either::EitherOutput,
+    },
+    identify::{
+        Behaviour as Identify,
+        Config as IdentifyConfig,
+        Event as IdentifyEvent,
+        Info as IdentifyInfo,
+    },
+    swarm::{
+        derive_prelude::{
+            ConnectionClosed,
+            ConnectionEstablished,
+            DialFailure,
+            FromSwarm,
+            ListenFailure,
+        },
+        ConnectionHandler,
+        IntoConnectionHandler,
+        IntoConnectionHandlerSelect,
+        NetworkBehaviour,
+        NetworkBehaviourAction,
+        PollParameters,
+    },
+    Multiaddr,
+    PeerId,
+};
+use tokio::time::{
+    self,
+    Interval,
+};
+
+use std::{
+    collections::VecDeque,
+    task::{
+        Context,
+        Poll,
+    },
+    time::Duration,
+};
+
+use tracing::debug;
+
+/// Maximum amount of peer's addresses that we are ready to store per peer
+const MAX_IDENTIFY_ADDRESSES: usize = 10;
+const HEALTH_CHECK_INTERVAL_IN_SECONDS: u64 = 10;
+const REPUTATION_DECAY_INTERVAL_IN_SECONDS: u64 = 1;
+
+/// Events emitted by PeerInfoBehaviour
+#[derive(Debug, Clone)]
+pub enum PeerReportEvent {
+    PeerConnected {
+        peer_id: PeerId,
+        addresses: Vec<Multiaddr>,
+        initial_connection: bool,
+    },
+    PeerDisconnected {
+        peer_id: PeerId,
+    },
+    PeerIdentified {
+        peer_id: PeerId,
+        agent_version: String,
+        addresses: Vec<Multiaddr>,
+    },
+    PeerInfoUpdated {
+        peer_id: PeerId,
+        block_height: BlockHeight,
+    },
+    /// Informs p2p service / PeerManager to check health of reserved nodes' connections
+    CheckReservedNodesHealth,
+    /// Informs p2p service / PeerManager to perform reputation decay of connected nodes
+    PerformDecay,
+}
+
+// `Behaviour` that reports events about peers
+pub struct PeerReportBehaviour {
```

Comment:
This seems a little confusing to call this peer report behavior, as it implies this service handles reputation. However, it seems like it handles nearly all the info for each peer.

Reply:
Good point, I mostly aimed at "peer reporting" as in report peer events, ie disconnected, connected, identified etc. I am definitely open to a better name